docs/proposals/config_data.md

```diff
@@ -0,0 +1,315 @@
+is:
+
+1. A `ConfigData` object is created
+2. A pod that consumes the configuration data via environment variables is created
+3. The pod is scheduled onto a node
+4. The kubelet retrieves the `ConfigData` resource(s) referenced by the pod and starts the container
+   processes with the appropriate data in environment variables
+
+### Consuming `ConfigData` in Volumes
+
+Many programs read their configuration from configuration files.  `ConfigData` should be possible
+to consume in a volume.  The rough series of events for consuming `ConfigData` this way
+is:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. The Kubelet creates an instance of the volume plugin and calls its `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+
+### Consuming `ConfigData`  Updates
+
+Any long-running system has configuration that is mutated over time.  Changes made to configuration
+data must be made visible to pods consuming data in volumes so that they can respond to those
+changes.
+
+The `resourceVersion` of the `ConfigData` object will be updated by the API server every time the
+object is modified.  After an update, modifications will be made visible to the consumer container:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. During the sync loop, the Kubelet creates an instance of the volume plugin and calls its
+   `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+6. The `ConfigData` referenced by the pod is updated
+7. During the next iteration of the `syncLoop`, the Kubelet creates an instance of the volume plugin
+   and calls its `Setup()` method
+8. The volume plugin projects the updated data into the volume
+
+It is the consuming pod's responsibility to make use of the updated data once it is made visible.
+
+Because environment variables cannot be updated without restarting a container, configuration data
+consumed in environment variables will not be updated.
+
+### Advantages
+
+* Easy to consume in pods; consumer-agnostic
+* Configuration data is persistent and versioned
+* Consumers of configuration data in volumes can respond to changes in the data
+
+## Proposed Design
+
+### API Resource
+
+The `ConfigData` resource will be added to the `extensions` API Group:
+
+```go
+package api
+
+// ConfigData holds configuration data for pods to consume.
+type ConfigData struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+  // Data contains the configuration data.  Each key must be a valid DNS_SUBDOMAIN or leading
+  // dot followed by valid DNS_SUBDOMAIN.
+	Data map[string]string `json:"data,omitempty"`
```

Comment:
I remember now why I didn't create that issue last night: we already have
an issue for this: #4789

On Thursday, November 19, 2015, Tim Hockin <notifications@github.com> wrote:

> In docs/proposals/config_data.md
> <https://github.com/kubernetes/kubernetes/pull/6477#discussion_r45416882>:
>
> > +
> > +### API Resource
> > +
> > +The `ConfigData` resource will be added to the `extensions` API Group:
> > +
> > +```go
> > +package api
> > +
> > +// ConfigData holds configuration data for pods to consume.
> > +type ConfigData struct {
> > +	TypeMeta   `json:",inline"`
> > +	ObjectMeta `json:"metadata,omitempty"`
> > +
> > +  // Data contains the configuration data.  Each key must be a valid DNS_SUBDOMAIN or leading
> > +  // dot followed by valid DNS_SUBDOMAIN.
> > +	Data map[string]string `json:"data,omitempty"`
>
> There's discussion in flight on the name restrictions for secrets. The
> absences of underscore, for example, really is an oversight.
>
> But mostly, the secrets volume is just wrong to assume that keys ==
> filenames.
>
> If we assume that few people has complained about the name format for
> secrets (yet), maybe we can rev the secrets volume to have an optional
> explicit file->key mapping, like downwardAPI. We still can't get rid of the
> leading dot without breaking compat, so I guess it's fine to leave this as
> the same spec. I just hope that doesn't bite us when we do some sort of
> templating for the volume for config data.
>
> So I rescind my comment. Making this the same spec as secrets is better.
> and we should fix secrets. @pmorie <https://github.com/pmorie>
> @derekwaynecarr <https://github.com/derekwaynecarr> - I will file a bug
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6477/files#r45416882>.
>